### PR TITLE
Support EndpointSlices in the controller/translator.

### PR DESCRIPTION
**Pull request by alan-kut**

### CODE REVIEW

File: pkg/controller/translator/translator.go [hunk 1]
```diff
@@ -549,7 +556,33 @@ func listAll(store cache.Store, selector labels.Selector, appendFn cache.AppendF
 	return nil
 }
 
-func listEndpointTargetPorts(indexer cache.Indexer, namespace, name, targetPort string) []int {
+func listEndpointTargetPortsFromEndpointSlices(indexer cache.Indexer, namespace, name, targetPort string) []int {
+	if i, err := strconv.Atoi(targetPort); err == nil {
```
Comment (freehan) [inline pkg/controller/translator/translator.go:560]: Add comment // if targetPort is integer, no need to translate to endpoint ports

File: pkg/controller/translator/translator.go [hunk 2]
```diff
@@ -549,7 +556,33 @@ func listAll(store cache.Store, selector labels.Selector, appendFn cache.AppendF
 	return nil
 }
 
-func listEndpointTargetPorts(indexer cache.Indexer, namespace, name, targetPort string) []int {
+func listEndpointTargetPortsFromEndpointSlices(indexer cache.Indexer, namespace, name, targetPort string) []int {
```
Comment (freehan) [inline pkg/controller/translator/translator.go:559]: add comment for both `listEndpointTargetPortsFromEndpointSlices` and `listEndpointTargetPortsFromEndpoints`:

return targetport if port number is specified. 

find the actual target port behind the named target port

File: pkg/controller/translator/translator.go [hunk 3]
```diff
@@ -549,7 +556,33 @@ func listAll(store cache.Store, selector labels.Selector, appendFn cache.AppendF
...
+	if err != nil {
+		klog.Errorf("Failed to retrieve endpoint slices for service %s/%s: %v", namespace, name, err)
+		return []int{}
+	}
+
+	ret := []int{}
+	for _, sliceObj := range slices {
+		slice := sliceObj.(*discoveryapi.EndpointSlice)
+		for _, port := range slice.Ports {
+			if port.Protocol != nil && *port.Protocol == api_v1.ProtocolTCP && port.Name != nil && *port.Name == targetPort && port.Port != nil {
```
Comment (freehan) [inline pkg/controller/translator/translator.go:577]: Looking at this piece of code. I realize the previous code has a bug in it. 

port.Name is actually the service port name. targetPort is the targetPort Name. They can be different. 

The correct implementation should be passing in service port name and look it up in the endpoint when targetport is named port.

Comment (freehan) [reply pkg/controller/translator/translator.go:577]: Might worth fixing this bug.  


See a counter example for the current implementation. 
https://github.com/freehan/kubescript/blob/master/hack/namdeport/namedport.yaml

**Comment by freehan** [comment]
So this PR from OSS community member was trying to fix the bug. #1462

This one adds a field ServicePortName field in ServicePort struct, which is missing for the translation.

File: pkg/controller/translator/translator.go (hunk 3 above)
Comment (alan-kut) [reply pkg/controller/translator/translator.go:577]: I'd leave fixing a bug into separate PR.

File: pkg/controller/translator/translator.go (hunk 1 above)
Comment (alan-kut) [reply pkg/controller/translator/translator.go:560]: Done

File: pkg/controller/translator/translator.go (hunk 2 above)
Comment (alan-kut) [reply pkg/controller/translator/translator.go:559]: Done

File: pkg/controller/translator/translator.go (hunk 3 above)
Comment (alan-kut) [reply pkg/controller/translator/translator.go:577]: The reasoning is that - fixing a bug and support Endpoint Slices are entirely different things. Having them separate would be more clear and safer.

**Review by freehan — APPROVED** [review]
/lgtm
/approve

**Pull request merged**